src/Agent.Plugins/PipelineCache/FingerprintCreator.cs

```diff
@@ -0,0 +1,238 @@
+                "system.defaultworkingdirectory" // Constants.Variables.System.DefaultWorkingDirectory
+                )?.Value;
+
+            var resolvedSegments = new List<string>();
+
+            foreach (string keySegment in keySegments)
+            {
+                if (keySegment.Equals("*", StringComparison.Ordinal))
+                {
+                    throw new ArgumentException("`*` is a reserved key segment. For path glob, use `./*`.");
+                }
+                else if (keySegment.Equals(Fingerprint.Wildcard, StringComparison.Ordinal))
+                {
+                    throw new ArgumentException("`**` is a reserved key segment. For path glob, use `./**`.");
+                }
+                else if (IsPathy(keySegment))
+                {
+                    context.Verbose($"Interpretting `{keySegment}` as a path.");
+
+                    var segment = new StringBuilder();
+                    bool foundFile = false;
+
+                    string[] pathRules = keySegment.Split(new []{','}, StringSplitOptions.RemoveEmptyEntries);
+                    string rootRule = pathRules.First();
+                    if(rootRule[0] == '!')
+                    {
+                        throw new ArgumentException("Path glob must start with an include glob.");
+                    }
+
+                    string workingDirectory = null;
+                    if (!Path.IsPathFullyQualified(rootRule))
+                    {
+                        workingDirectory = workingDirectoryValue;
+                    }
+
+                    string absoluteRootRule = MakePathAbsolute(workingDirectory, rootRule);
+                    context.Verbose($"Expanded include rule is `{absoluteRootRule}`.");
+                    IEnumerable<string> absoluteExcludeRules = pathRules.Skip(1).Select(r => {
+                        if (r[0] != '!')
+                        {
+                            throw new ArgumentException("Path glob must start with an exclude glob.");
+                        }
+                        return MakePathAbsolute(workingDirectory, r.Substring(1));
+                    });
+                    Func<string,bool> filter = CreateFilter(context, workingDirectory, absoluteRootRule, absoluteExcludeRules);
+
+                    DetermineFileEnumerationFromGlob(
+                        absoluteRootRule,
+                        out string enumerateRootPath,
+                        out string enumeratePattern,
+                        out SearchOption enumerateDepth);
+
+                    context.Verbose($"Enumerating starting at root `{enumerateRootPath}` with pattern `{enumeratePattern}`.");
+                    IEnumerable<string> files = Directory.EnumerateFiles(enumerateRootPath, enumeratePattern, enumerateDepth);
+                    files = files.Where(f => filter(f)).Distinct();
+
+                    foreach(string path in files)
```

Comment:
May want to sort this list (by name) in case `EnumerateFiles` (or any of the filtering logic) doesn't return files in a consistent order. Tos say it another way, the order of the matching files for a segment shouldn't impact the fingerprint.

Reply:
We'll hash in enumeration order and then create the key segment in path order